File: day-5a/src/main.py
import dataclasses as dc
import enum
from collections.abc import Iterator
from functools import cached_property
from typing import Self

class GardeningThing(enum.Enum):
    seed = enum.auto()
    soil = enum.auto()
    fertilizer = enum.auto()
    water = enum.auto()
    light = enum.auto()
    temperature = enum.auto()
    humidity = enum.auto()
    location = enum.auto()


@dc.dataclass(kw_only=True, slots=True, frozen=True)
class MapKind:
    source: GardeningThing
    destination: GardeningThing

    def __repr__(self) -> str:
        return f"MapKind({self.source.name}-to-{self.destination.name})"
    
    @classmethod
    def from_description(cls, description: str) -> Self:
        source, _, destination = description.partition("-to-")
        return cls(
            source=GardeningThing[source],
            destination=GardeningThing[destination]
        )


@dc.dataclass(frozen=True)
class InputDataRow:
    destination_start: int
    source_start: int
    range_length: int

    @cached_property
    def source_range(self) -> range:
        return range(self.source_start, self.source_start + self.range_length)


@dc.dataclass(slots=True, kw_only=True, frozen=True)
class Map:
    kind: MapKind
    rows: tuple[InputDataRow, ...] = dc.field(repr=False)

    def convert(self, item: int) -> int:
        for row in self.rows:
            if item in row.source_range:
                difference = item - row.source_start
                return row.destination_start + difference
        else:
            return item


def location_from_seed(seed: int, maps: tuple[Map, ...]) -> int:
    answer = seed
    thing = GardeningThing.seed
    while thing is not GardeningThing.location:
        relevant_map = next(m for m in maps if m.kind.source is thing)
        answer = relevant_map.convert(answer)
        thing = relevant_map.kind.destination
    return answer


@dc.dataclass(slots=True, kw_only=True, frozen=True)
class InputData:
    seeds: frozenset[int]
    maps: tuple[Map, ...]

    def seed_locations(self) -> Iterator[int]:
        for seed in self.seeds:
            yield location_from_seed(seed, self.maps)


def parse_input(filename: str) -> InputData:
    with open(filename) as f:
        puzzle_input = f.read()
    unparsed_seeds, *unparsed_maps = puzzle_input.split("\n\n")
    seeds = frozenset(map(int, unparsed_seeds.split(" ")[1:]))
    maps: list[Map] = []
    for unparsed_map in unparsed_maps:
        first_line, *rest = unparsed_map.splitlines()
        kind = MapKind.from_description(first_line.partition(" ")[0])
        rows = tuple(InputDataRow(*map(int, line.split())) for line in rest)
        maps.append(Map(kind=kind, rows=rows))
    return InputData(seeds=seeds, maps=tuple(maps))


def solve(filename: str) -> int:
    input_data = parse_input(filename)
    return min(input_data.seed_locations())


def main() -> None:
    print(solve("input.txt"))


if __name__ == "__main__":
    main()

File: day-5a/src/main.rs
use std::{fs::read_to_string, ops::Range};

use cached::proc_macro::cached;

#[derive(PartialEq, Clone)]
enum GardeningThing {
    Seed,
    Soil,
    Fertilizer,
    Water,
    Light,
    Temperature,
    Humidity,
    Location,
}

#[cached]
fn gardening_thing_from_description(description: String) -> GardeningThing {
    match description.as_str() {
        "seed" => GardeningThing::Seed,
        "soil" => GardeningThing::Soil,
        "fertilizer" => GardeningThing::Fertilizer,
        "water" => GardeningThing::Water,
        "light" => GardeningThing::Light,
        "temperature" => GardeningThing::Temperature,
        "humidity" => GardeningThing::Humidity,
        "location" => GardeningThing::Location,
        _ => panic!(),
    }
}

struct MapKind {
    source: GardeningThing,
    destination: GardeningThing,
}

impl MapKind {
    fn from_descriptions(source_description: &str, destination_description: &str) -> MapKind {
        MapKind {
            source: gardening_thing_from_description(String::from(source_description)),
            destination: gardening_thing_from_description(String::from(destination_description)),
        }
    }
}

struct InputDataRow {
    destination_start: u32,
    source_start: u32,
    range_length: u32,
}

impl InputDataRow {
    fn source_range(&self) -> Range<u32> {
        self.source_start..(self.source_start.wrapping_add(self.range_length))
    }
}

struct Map {
    kind: MapKind,
    rows: Vec<InputDataRow>,
}

impl Map {
    fn convert(&self, item: u32) -> u32 {
        for row in &self.rows {
            if row.source_range().contains(&item) {
                let difference = item - row.source_start;
                return row.destination_start + difference;
            }
        }
        item
    }
}

fn location_from_seed(seed: u32, maps: &Vec<Map>) -> u32 {
    let mut answer = seed;
    let mut thing = &GardeningThing::Seed;
    while thing != &GardeningThing::Location {
        let relevant_map = maps
            .iter()
            .filter(|m| &m.kind.source == thing)
            .next()
            .unwrap();
        answer = relevant_map.convert(answer);
        thing = &relevant_map.kind.destination;
    }
    answer
}

struct InputData {
    seeds: Vec<u32>,
    maps: Vec<Map>,
}

impl InputData {
    fn seed_locations(&self) -> impl Iterator<Item = u32> + '_ {
        self.seeds
            .iter()
            .map(|s| location_from_seed(*s, &self.maps))
    }
}

fn parse_row_from_input(unparsed_row: &str) -> InputDataRow {
    match unparsed_row
        .split_whitespace()
        .into_iter()
        .map(|s| s.parse::<u32>().unwrap())
        .collect::<Vec<u32>>()[..]
    {
        [destination_start, source_start, range_length] => InputDataRow {
            destination_start,
            source_start,
            range_length,
        },
        _ => panic!(),
    }
}

fn parse_kind_from_input(kind_description: &str) -> MapKind {
    match kind_description.split("-").collect::<Vec<&str>>()[..] {
        [source_description, _, destination_description] => {
            MapKind::from_descriptions(source_description, destination_description)
        }
        _ => panic!(),
    }
}

fn parse_map_from_input(unparsed_map: &str) -> Map {
    match &unparsed_map.split("\r\n").collect::<Vec<&str>>()[..] {
        [first_line, unparsed_rows @ ..] => {
            assert!(unparsed_rows.len() > 1);
            let kind_description = first_line.split(" ").next().unwrap();
            let kind = parse_kind_from_input(kind_description);
            let mut rows = Vec::<InputDataRow>::new();
            for unparsed_row in unparsed_rows {
                rows.push(parse_row_from_input(&unparsed_row))
            }
            Map { kind, rows }
        }
        _ => panic!(),
    }
}

fn parse_seeds_from_input(seed_description: &str) -> Vec<u32> {
    seed_description.split(" ").collect::<Vec<&str>>()[1..]
        .iter()
        .map(|s| s.parse::<u32>().unwrap())
        .collect::<Vec<u32>>()
}

fn parse_input(filename: &str) -> InputData {
    let puzzle_input = read_to_string(filename).unwrap();
    let mut maps = Vec::<Map>::new();
    let seeds = match &puzzle_input.split("\r\n\r\n").collect::<Vec<&str>>()[..] {
        [unparsed_seeds, unparsed_maps @ ..] => {
            assert!(unparsed_maps.len() > 1);
            let seeds = parse_seeds_from_input(&unparsed_seeds);
            for unparsed_map in unparsed_maps {
                maps.push(parse_map_from_input(unparsed_map))
            }
            seeds
        }
        _ => panic!(),
    };
    InputData { seeds, maps }
}

fn solve(filename: &str) -> u32 {
    let input_data = parse_input(filename);
    input_data.seed_locations().min().unwrap()
}

fn main() {
    println!("{}", solve("input.txt"));
}
